### PR TITLE
feat: Implement TopicConfigWatcher

### PR DESCRIPTION
**Pull request by palmere-google**
The topic config watcher polls the topic config and calls an update handler whenever the config cahnges

Fixes #<issue_number_goes_here> ☕️

### CODE REVIEW

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcherImpl.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+public class TopicConfigWatcherImpl extends AbstractApiService implements TopicConfigWatcher {
+  private static final GoogleLogger log = GoogleLogger.forEnclosingClass();
+  private final Duration period;
+  private final TopicPath topicPath;
+  private final AdminClient adminClient;
+  private final ScheduledExecutorService executorService;
+  private final Consumer<Topic> configReceiver;
+
+  private ScheduledFuture<?> topicConfigPoll;
+  private Topic currentConfig = null;
```
Comment (dpcollins-google) [inline google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcherImpl.java:37]: Use Optional\<Topic\>?

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcher.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+package com.google.cloud.pubsublite.internal.wire;
+
+import com.google.api.core.ApiService;
+import com.google.cloud.pubsublite.proto.Topic;
+import java.time.Duration;
+import java.util.function.Consumer;
+
+public interface TopicConfigWatcher extends ApiService {
```
Comment (dpcollins-google) [inline google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcher.java:23]: I don't think Topic is the correct level of abstraction. Publishers do not need GetTopic permissions to run: you should make this a TopicPartitionsCountWatcher.

**Review by dpcollins-google — CHANGES_REQUESTED** [review]

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcher.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  interface Factory {
+    TopicConfigWatcher New(Consumer<Topic> receiver, Duration period);
```
[the remark this answers is not in the document]
Comment (palmere-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcher.java:25]: I'm confused, what do you mean by not valid? We have other classed with a New method

e.g. https://github.com/googleapis/java-pubsublite/blob/8c7ebd0acf1c1677b679f67e233e49495b889ea5/google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/AssignerFactory.java#L22

(I don't particularly object to changing it)

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcher.java (hunk 2 above)
Comment (palmere-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcher.java]: I completely forgot that we had a separate rpc for fetching the partition count. Changed

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcherImpl.java (hunk 1 above)
Comment (palmere-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcherImpl.java]: Yeah, that's better. Done

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcher.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcher.java]: I think the linter complains about it but can't automatically fix it :/

[the remark this answers is not in the document]
Comment (palmere-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/TopicConfigWatcher.java]: Ah, okay, changed to newWatcher

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/PartitionCountWatcher.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package com.google.cloud.pubsublite.internal.wire;
+
+import com.google.api.core.ApiService;
+import java.time.Duration;
+import java.util.function.Consumer;
+
+public interface PartitionCountWatcher extends ApiService {
+  interface Factory {
+    PartitionCountWatcher New(Consumer<Long> receiver, Duration period);
```
Comment (dpcollins-google) [inline google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/PartitionCountWatcher.java:24]: The duration should not be a factory parameter.

Comment (palmere-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/PartitionCountWatcher.java]: Done

**Review by dpcollins-google — APPROVED** [review]

**Pull request merged**